Add Two Preprocessing Options

hmc-cs-mdrissi opened the pull request:
The preprocessing options added are an option to choose which tokenization algorithm used (current choices being the old one as default and nltk's word_tokenize) and a max_length for both source/target. The max_length just truncates any sequences as needed. If no max_length is desired than -1 works and is the default.

alexeib commented on preprocess.py:
```diff
@@ -40,18 +41,30 @@ def get_parser():
+                        help="The maximum length of the source. If a sequence is longer it will be truncated to this length. " 
+                             "If -1, then no truncation occurs. For joined_dictionary's the maximum of the source/target max length will"
+                             " be used to build the dictionary. Default is no truncation.")
+    parser.add_argument('--max_target_length', metavar='N', default=-1, type=int,
```
can you just remove the defaults and check if max_source_length and max_target_length are None instead (instead of -1)

alexeib commented on fairseq/tokenizer.py:
```diff
@@ -50,6 +55,7 @@ def replaced_consumer(word, idx):
                     consumer=replaced_consumer,
                     append_eos=append_eos,
                     reverse_order=reverse_order,
+		    max_length=max_length
```
indentation seems to be off and missing trailing comma

alexeib commented on preprocess.py:
```diff
@@ -40,18 +41,30 @@ def get_parser():
     parser.add_argument('--padding-factor', metavar='N', default=8, type=int,
                         help='Pad dictionary size to be multiple of N')
+    parser.add_argument('--tokenize_number', metavar='N', default=0, type=int, choices=range(2),
+                        help="Which tokenization algorithm to use. 0 is tokenize_line, 1 is nltk's word tokenize. Default is 0.")
```
it is better to use string arguments here with a choices=['default', 'nltk'], here and then store them in a dict instead of array. also can you please describe what the difference between the two tokenization strategies are, and add a summary of that description into the help string?

hmc-cs-mdrissi commented:
My main motivation for wanting to limit the lengths of inputs/targets during preprocessing is to limit the number of tokens in a single sequence. For the stories data set some of the target stories are over 1000 words long. The paper originally truncated those stories to 1000 words. While I could change max_target_positions that would lead to training/validating on the entire sequence which isn't intended.

myleott commented:
Supporting alternative tokenizers seems useful, e.g., it'd be cool to support SacreMoses too: https://github.com/alvations/sacremoses.

Let's restructure this a bit though, because I'm reluctant to introduce a dependency on nltk, and think it'd be more generally useful to make an NLTKTokenizer class that extends the base Tokenizer.

Something like:
```
class NLTKTokenizer(Tokenizer):

  def __init__(self):
    try:
      from nltk import word_tokenize
    except ImportError as e:
      import sys
      sys.stderr.write('ERROR: please install nltk to use ')
      raise e

    # punkt is necessary to use word_tokenize.
    if not nltk.downloader.Downloader().is_installed('punkt'):
      nltk.download('punkt')

  def tokenize(self, line, dict, tokenize=tokenize_line, add_if_not_exist=True,
               consumer=None, append_eos=True, reverse_order=False):
      (...)
```

We should also introduce a `build_tokenizer` method that abstracts the tokenizer selection, similar to what we do for progress_bars: https://github.com/pytorch/fairseq/blob/master/fairseq/progress_bar.py#L22-L39. Along these lines, we should replace all the static calls into Tokenizer with a call to build_tokenizer and then method calls, e.g., in score.py, IndexedDataset, generate.py.

Lastly, I'm not sure I understand why `--max-target-positions` doesn't accomplish your goal. Filtering based on length in preprocessing seems error prone; the length checks during validation are important to make sure you're not inadvertently biasing your validation/test evaluation. The `--skip-invalid-size-inputs-valid-test` is there specifically to circumvent this, if needed.

hmc-cs-mdrissi commented:
I don't want to filter out long inputs entirely. I only want to truncate them (mainly for consistency with how the data set was used in the past). Currently if I set max-target-positions at 1000 without doing any dataset modification either I get an error (if skip-invalid is not used) or I completely skip those examples. 

As for why not truncate during train/test time that would change the vocab size. Given a dataset with long sequences, if you truncate at preprocessing than words after the truncation point will not be a part of the vocab. Truncating during train time keeps all of those words even though they should never be predicted.

I'll make the tokenizer changes later today/tomorrow.

hmc-cs-mdrissi commented:
For restructing the tokenizer.py, I've added a build_tokenizer, made the Tokenizer's functions all no longer static, and added NLTK/SacreMoses as subclasses (they only override the init). I cleaned up a bit how max_length was passed around so that it's taken care of when the tokenizer is built.

While I've added tokenizer_name as an option everywhere I didn't add max_length to them. I can add that to any of them if desired. Max_length could technically appear for all the parsers as other parsers like for train/eval_lm also needed tokenizer_name to account for the tokenization that might happen when load_dataset is called (it happens when the raw_text option is used). If I were to add it I'd likely add it to the two task files so that one of them would get a max_source/target_length while the other would just have a max_length and have them take responsibility for building the tokenizer (the current set up where it's built at times in them and at times in the main file being called feels a bit messy). As every single parser made in the options file needed it (and will continue to as long as all the tasks might use it), I ended up adding it as an optional argument in get_parser. 

One other thing is none of the current tokenizers have hyperparameters in them. If a tokenization algorithm is ever added that does (like byte pair) then it's options will need to likely be dealt with in a way similar to how model options are dealt with.

One assumption I've gone with is that the tokenization algorithm used on the source/target text is the same. If we'd prefer making it possible for them to use different tokenization algorithms, I could later break the option tokenizer_name into two options as needed. This does have some messiness associated to it because files that could be used for either task telling whether one/two tokenizers is needed I'll need to ponder. Similar to if I had to add max_length as an option everywhere I'd likely end up having to make the tasks take care of building tokenizers. It might be easiest to pretend that both tasks have a source/target tokenizer_tool and then in the case of language models the two tokenizers are equal. I'd also need to decide what happens if you choose different tokenization algorithms and a joint dictionary (I think this should just throw an error).

Lastly, the name tokenizer_tool used for the built tokenizer instead of the simpler name of just tokenizer was because I didn't want to shadow the module name. Also all the unit tests pass.

hmc-cs-mdrissi commented:
@alexeib, @myleott: Review/comments?

myleott commented:
Sorry for the delayed review, just got back from ICML. I made some initial comments, but looks pretty good otherwise.

hmc-cs-mdrissi commented:
@myleott I don't see the initial comments at the moment? Did you forget to submit them?
